### PR TITLE
add an integration test for muxer selection

### LINKED ISSUE #1836: Create integration test for libp2p early muxer negotiation
The code changes that are specified in [#446] changes muxer negotiation behavior, an integration test plan should be put in  place to test the correct muxer negotiation behavior.

[#446]:  https://github.com/libp2p/specs/pull/446/files

### PR DESCRIPTION
**Pull request by marten-seemann**
Fixes #1836. Depends on #1885. Closes #1851. Closes #1840.

The test currently fails since apparently the early muxer negotiation respects the server's preferences, not the client's. We should fix that. Will do in another PR.

### CODE REVIEW

**Comment by MarcoPolo** [comment]
fyi I moved the base to `remove-muxer-multistream`

File: p2p/net/upgrader/upgrader.go [hunk 1]
```diff
@@ -257,17 +263,18 @@ func (u *upgrader) setupMuxer(ctx context.Context, conn sec.SecureConn, server b
 		if err != nil {
 			return
 		}
+		muxerID = m.ID
 		smconn, err = m.Muxer.NewConn(conn, server, scope)
 	}()
 
 	select {
 	case <-done:
```
Comment (MarcoPolo) [inline p2p/net/upgrader/upgrader.go:271]: Minor nit, I think it would be better to return the value here (a la message passing) rather than mutating shared state.

File: config/config.go [hunk 2]
```diff
@@ -182,6 +182,7 @@ func (cfg *Config) addTransports(h host.Host) error {
 			fx.ParamTags(`group:"security"`),
 		)),
 		fx.Supply(cfg.Muxers),
+		fx.Supply(h.ID()),
```
Comment (MarcoPolo) [inline config/config.go:185]: Why this change? Did this not work before?

**Review by MarcoPolo — APPROVED** [review]
lg, one nit.

File: config/config.go (hunk 2 above)
Comment (marten-seemann) [reply config/config.go:185]: The insecure transport consume the `peer.ID`: https://github.com/libp2p/go-libp2p/blob/d8d2efafdd368bc4911c734ee284ff285fcdfbca/core/sec/insecure/insecure.go#L43

The other transports consume the private key.

**Comment by marten-seemann** [comment]
Rebased and changed merge target back to `master`.

File: p2p/net/upgrader/upgrader.go (hunk 1 above)
Comment (marten-seemann) [reply p2p/net/upgrader/upgrader.go]: I think I agree. Changed.
